[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations

george-qi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84887
* __->__ #82841

Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

george-qi committed:
[maskedtensor] add more basic tests

[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add more basic tests"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add more basic tests"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add more basic tests"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
 
 
+class TestBasics(TestCase):
+    def test_add(self):
```
this is a negative test of (I think) incompatible masks, name needs to reflect what is being tested. also idk if this means add() isn't being positive tested but if not that should be added

george-qi replied:
added the positive test!

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
+        ts = torch.softmax(mx, -1)
+        ts.sum().backward()
+        xinf = x.masked_fill(~m, float("-inf")).detach().clone().requires_grad_()
+        torch.softmax(xinf, -1)
```
what is being tested here

george-qi replied:
added an actual test for equivalence

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
+        my = MaskedTensor(torch.ones_like(x), ~mask, requires_grad=True)
+        y = torch.where(mask, torch.exp(mx), my)
+        s = y.sum()
+        s.backward()
```
what is being tested here

george-qi replied:
added an equivalence test

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
+        s = y.sum()
+        s.backward()
+
+    def test_mha_issue_41508(self):
```
this isn't a sufficient test of the MT `multi_head_attention_forward()`. we need to either fully test it a la the core, non-MT version, or not ship it.

the fact that this is the only test makes me think that its purpose is basically a demo of how MTs can be used to fix 41508. That's cool but it doesn't motivate adding it to core - throwing yet another mha function in there is counterproductive, it just confuses things wrt what mha function to use. Personally I'd say we should remove it but @albanD @cpuhrsch might have other opinions.

george-qi replied:
removed for now as per our offline discussion

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
+        w = torch.tensor([1.0, 2.0], requires_grad=True)
+        w_q, w_k = w.chunk(2)
+        o0 = k + w_k
+        o0.backward()
```
what is being tested

george-qi replied:
removed

bhosmer reviewed (commented):
quick first round of comments

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

cpuhrsch commented on test/test_maskedtensor.py:
```diff
@@ -281,84 +439,84 @@ def test_mean(self):
-            MaskedTensor(data, mask, requires_grad=True/False) -- NO differentiable constructor and always a leaf
-            MaskedTensor.from_values(data, mask) -- differentiable constructor
+            masked_tensor(data, mask, requires_grad=True/False) -- NO differentiable constructor and always a leaf
+            masked_tensor(data, mask) -- differentiable constructor
```
Did you mean `as_masked_tensor` as in the piece of text below? Please check for this error within the PR and ask for review once you're confident it's been resolved.

george-qi replied:
Yeah thanks, fixed

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"




[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`MaskedTensor()` -> `masked_tensor()`
`MaskedTensor.from_values()` -> `as_masked_tensor()`


[ghstack-poisoned]

cpuhrsch commented on test/test_maskedtensor.py:
```diff
@@ -536,24 +692,154 @@ def test_masked_bmm(self):
-        v = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
-        v.sum().backward()
 
-    def test_linear(self):
```
What happened to this test?

george-qi replied:
It wasn't actually testing an op on MaskedTensor afaik; it just performed `torch.nn.functional.linear` on a regular Tensor and then put it inside a `MaskedTensor`. Unless you had another test in mind to add?

cpuhrsch replied:
Does MaskedTensor support linear? If so we might need to add the test.

cpuhrsch commented on torch/_masked/__init__.py:
```diff
@@ -1067,19 +1067,19 @@ def sum(
     dim_ = _canonical_dim(dim, input.ndim)
     mask_input = _combine_input_and_mask(sum, input, mask)
-    if input.layout == torch.strided:
+    if mask_input.layout == torch.strided:
```
Why change this?

george-qi replied:
This was changed because the code was updated to support MaskedTensor inputs (in particular, if the input is MaskedTensor, then we want a differentiable `_combine_input_and_mask`, but we don't need one for vanilla tensor/masks. And so it's more accurate to check the `mask_input`'s layout, which is post-processing, rather than the input

cpuhrsch commented on torch/masked/maskedtensor/binary.py:
```diff
@@ -94,7 +94,10 @@ def _binary_helper(fn, args, kwargs, inplace):
     args0_layout = data_args[0].layout
-    same_layout = args0_layout == data_args[1].layout
+    same_layout = (
+        (torch.is_tensor(data_args[1]) or is_masked_tensor(data_args[1])) and
```
Why is this additional check necessary?

george-qi replied:
This is necessary because there are cases where we do `MaskedTensor + float`, and `float`s don't have a dtype so that would error out

cpuhrsch commented on torch/masked/maskedtensor/core.py:
```diff
@@ -342,7 +342,7 @@ def from_values(data, mask):
             @staticmethod
             def forward(ctx, data, mask):
-                return MaskedTensor(data.clone(), mask.clone())
+                return MaskedTensor(data, mask)
```
Do we want to keep around `from_values`?

george-qi replied:
I'm okay either way -- I suppose we could remove it, but I suppose also no harm. I suppose the more disciplined way would just have `masked_tensor` call `MaskedTensor()` and then `as_masked_tensor` call `MaskedTensor.from_values` -- I'll change it to this if we're all okay with that?

cpuhrsch commented on torch/masked/maskedtensor/reductions.py:
```diff
@@ -45,30 +46,35 @@ def _get_masked_fn(fn):
 def _torch_reduce_all(fn):
```
As a general note on maskedtensor reductions, there already is a lot of code in torch/_masked. What are the plans for that code?

george-qi replied:
Discussed offline -- we'll move this into `torch/masked`

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`MaskedTensor()` -> `masked_tensor()`
`MaskedTensor.from_values()` -> `as_masked_tensor()`


[ghstack-poisoned]

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,155 @@ def _fix_fn_name(fn_name):
+        tensor_res = torch.softmax(xinf, -1)
+        tensor_res.sum().backward()
+        _compare_mt_t(masked_res, tensor_res)
+        if mt.grad is not None and xinf.grad is not None:
```
not sure I get the reason for the conditional here

george-qi replied:
I think was an old check when the entire mask was masked out to be safe even though the entries should just be `nan`. Regardless, not necessary now for sure. Fixed

bhosmer commented on test/test_maskedtensor.py:
```diff
@@ -96,6 +105,155 @@ def _fix_fn_name(fn_name):
+        tensor_res.sum().backward()
+
+        _compare_mt_t(mx.grad, x.grad)
+        _compare_mt_t(my.grad, y.grad)
```
are we testing the primal here?

george-qi replied:
Makes sense to test as well, added

bhosmer reviewed (commented):
quick comment pass, noted a couple basic things, will pick it up tomorrow

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

bhosmer reviewed (commented):
Couple more basic issues but looks like we're getting close.

High level question: `TestBasics` are kind of spot checks, while `TestOperators` does the more thorough sweeps using `torch.testing` infra . But afaict only `TestBasics` does backward() testing. Is that typical?

george-qi commented on test/test_maskedtensor.py:
```diff
@@ -536,24 +697,154 @@ def test_masked_bmm(self):
-            x, ~(key_padding_mask.transpose(0, 1).unsqueeze(-1).expand_as(x))
-        )
+        mt_res = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
+        mt_res.sum().backward()  # make sure this runs
```
I've been thinking about this some more, and I might hav some fundamental misunderstanding here, but I don't believe the gradients should match.. please correct me if I misunderstood something cc @cpuhrsch since you originally authored `MaskedBmm` so likely know it better than I:

Consider the `MaskedBmm` implementation:

https://github.com/pytorch/pytorch/blob/3b9486461eac463771e45d6ecc0b3459cb199562/torch/masked/maskedtensor/matmul.py#L15-L41

and the derivative from `derivatives.yaml`:

https://github.com/pytorch/pytorch/blob/33352336b443dbfce1394f6b4950e8f33eff2cef/tools/autograd/derivatives.yaml#L348-L351

Indeed, `q_grad = torch.bmm(grad, k_transpose)` and `k_grad = torch.bmm(q_transpose, grad)`

In this testing, we've verified that `mt_res` is equivalent to `t_res` -- that is, the forward passes are equivalent, because we've masked out `k` when we call `x_masked = x.masked_fill(~x_mt.get_mask(), 0).detach().clone().requires_grad_()`

However, grad is not masked out in the backwards pass of vanilla `torch.bmm`, whereas it is implicitly in line 38 of the `MaskedBmm` implementation. 

In addition, instead of `torch.bmm(grad_data, k_trans)` (line 35 above) it should probably be `torch.bmm(grad.to_tensor(0), k_trans` -- is that right?

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

george-qi commented on test/test_maskedtensor.py:
```diff
@@ -536,24 +697,154 @@ def test_masked_bmm(self):
+
+            # If the operation is binary, check that lhs = masked, rhs = regular tensor also works
+            if is_binary(op):
+                mt_result2 = op(mt, *mt_args, **sample_kwargs)
```
Ugh sorry this was stale, thanks for the catch. Updated this; this only works when `torch.layout == torch.strided`, and of course `mt_args` should be `sample_args`.

george-qi commented:
> Couple more basic issues but looks like we're getting close.
> 
> High level question: `TestBasics` are kind of spot checks, while `TestOperators` does the more thorough sweeps using `torch.testing` infra . But afaict only `TestBasics` does backward() testing. Is that typical?

Not entirely sure what's typical -- Christian and I have had more adhoc testing along these lines in the out-of-tree repoi, but I suppose part of migrating into core is adjusting to core's stricter testing, which is definitely a good thing :)

But as of this comment (because of your previous comment), I think every function in `TestBasics` had been changed to also testing the forward testing? e.g.:

https://github.com/pytorch/pytorch/blob/ea6c9c1131ae1282f95b89133504b9cef50d48bc/test/test_maskedtensor.py#L151

https://github.com/pytorch/pytorch/blob/ea6c9c1131ae1282f95b89133504b9cef50d48bc/test/test_maskedtensor.py#L220

and so on

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

cpuhrsch commented:
@pytorchbot rebase

george-qi committed:
Update on "[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations"


Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

[ghstack-poisoned]

george-qi commented:
@pytorchbot merge
